File: packages/observability/src/retry.ts
// Retry wrapper for external provider calls (LLM/STT/TTS). Mirrors
// apps/agent/agent/retry.py. Retries TransientProviderError with exponential
// backoff; PermanentProviderError is never retried.

export class TransientProviderError extends Error {}
export class PermanentProviderError extends Error {}

export interface RetryOptions {
  maxAttempts?: number;
  backoff?: "exponential";
  baseDelayMs?: number;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export async function withRetry<T>(
  fn: () => Promise<T>,
  opts: RetryOptions = {},
): Promise<T> {
  const { maxAttempts = 3, baseDelayMs = 200 } = opts;
  let attempt = 0;
  for (;;) {
    try {
      return await fn();
    } catch (err) {
      if (err instanceof PermanentProviderError) throw err;
      if (!(err instanceof TransientProviderError)) throw err;
      attempt += 1;
      if (attempt >= maxAttempts) throw err;
      await sleep(baseDelayMs * 2 ** (attempt - 1));
    }
  }
}

File: apps/agent/agent/retry.py
"""Retry + backoff wrapper for external provider calls (LLM/STT/TTS).

Mirrors packages/observability withRetry (TS). Retries TransientProviderError
with exponential backoff up to max_attempts; PermanentProviderError is never
retried.
"""
from __future__ import annotations

import time
from typing import Callable, TypeVar

T = TypeVar("T")


class TransientProviderError(Exception):
    """Retryable provider failure (rate limit, timeout, transient 5xx)."""


class PermanentProviderError(Exception):
    """Non-retryable provider failure (auth failure, hard limit, bad request)."""


def with_retry(
    fn: Callable[[], T],
    *,
    max_attempts: int = 3,
    base_delay: float = 0.2,
    sleep: Callable[[float], None] = time.sleep,
) -> T:
    attempt = 0
    while True:
        try:
            return fn()
        except PermanentProviderError:
            raise
        except TransientProviderError:
            attempt += 1
            if attempt >= max_attempts:
                raise
            sleep(base_delay * (2 ** (attempt - 1)))
